### PR TITLE
add GPTJ/bloom/llama/opt into model list and enhance the jit support

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

extend the text generation to more model

- generate: @gante

### CODE REVIEW

**Comment by sywangyi** [comment]
@amyeroberts please help review

**Comment by sywangyi** [comment]
@jiqing-feng @yao-matrix

**Review by gante — APPROVED** [review]
LGTM, thank you for the contribution :)

**Comment by gante** [comment]
@amyeroberts for context, this PR build upon https://github.com/huggingface/transformers/pull/22265 -- an example of how to JIT trace text generation

File: examples/pytorch/text-generation/run_generation.py [hunk 1]
```diff
@@ -238,15 +249,13 @@ def __init__(self, optimized, default):
         self._default = default
 
     def __call__(self, *args, **kwargs):
-        if kwargs["past_key_values"] is None:
-            return self._default(*args, **kwargs)
-        trace_graph_inputs = []
+        if kwargs["past_key_values"] is None and self._default.config.use_cache:
+            kwargs["past_key_values"] = generate_past_key_values(self._default, kwargs["input_ids"].shape[0], 0)
```
Comment (amyeroberts) [inline examples/pytorch/text-generation/run_generation.py:253]: Is this right? For the jit compilation, it's OK to use empty tensors, but in this case, wouldn't we be passing in tensors of random values as the past key, value values?

File: examples/pytorch/text-generation/run_generation.py [hunk 2]
```diff
@@ -173,61 +186,59 @@ def sparse_model_config(model_config):
...
+                .to(model.dtype)
                 .to(model.device),
             )
             for _ in range(num_block_layers)
         )
+    return past_key_values
 
+
+def prepare_jit_inputs(inputs, model, tokenizer):
+    num_batch = len(inputs)
```
Comment (amyeroberts) [inline examples/pytorch/text-generation/run_generation.py:227]: `batch_size` is better here - `num_batch` indicates the number of minibatches rather than the size of each batch

```suggestion
    batch_size = len(inputs)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work on adding this! 

In general, I'm not a huge fan changes like these. The examples should be demonstrative of the general use case of the library, they're not meant to be exhaustive or work for all models or settings: it's more important to be easier to read and adaptable for users. 

However, LLama and OPT are good to have support for and this addition has lead to some nice refactoring with `generate_past_key_values` and `prepare_jit_inputs`. 

Just a few comments to address and I think we're good to merge.

File: examples/pytorch/text-generation/run_generation.py [hunk 3]
```diff
@@ -382,14 +394,25 @@ def main():
...
+        else:
+            if hasattr(model, "forward"):
+                sig = inspect.signature(model.forward)
+            else:
+                sig = inspect.signature(model.__call__)
+            jit_inputs = tuple(jit_inputs[key] for key in sig.parameters if jit_inputs.get(key, None) is not None)
+            traced_model = torch.jit.trace(model, jit_inputs, strict=False)
+            traced_model = torch.jit.freeze(traced_model.eval())
+            traced_model(*jit_inputs)
+            traced_model(*jit_inputs)
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/pytorch/text-generation/run_generation.py:415]: Hi, @amyeroberts , thanks for the review,  I have a fix for jit.trace with dict input which needs be used in text generation case. the fix is https://github.com/pytorch/pytorch/pull/97866, contained in pytorch nightly version(2.1.0.xxx). 2.0 version does not have this fix. Without the fix, we could not use jit.trace with dict input for text generation. it will fail caused by the past_key_value. As for tuple input for jit.trace. it has some limitation, such as there should not be any parameter appeared in the middle of jit_inputs variables in the forward() func, like LlamaForCausalLM
position_id in forward(), appeared before past_key_value. which will make the jit.trace confused and fail. in the nightly pytorch version. jit.trace for mostly model will success with the dict input. That's why I add version check here.

File: examples/pytorch/text-generation/run_generation.py [hunk 4]
```diff
@@ -382,14 +394,25 @@ def main():
...
-        jit_input_texts = ["jit"]
+        jit_input_texts = ["enable jit"]
         jit_inputs = prepare_jit_inputs(jit_input_texts, model, tokenizer)
         torch._C._jit_set_texpr_fuser_enabled(False)
         model.config.return_dict = False
-        traced_model = torch.jit.trace(model, jit_inputs, strict=False)
-        traced_model = torch.jit.freeze(traced_model.eval())
-        traced_model(*jit_inputs)
-        traced_model(*jit_inputs)
+        if version.parse(version.parse(torch.__version__).base_version) >= version.parse("2.1.0"):
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/pytorch/text-generation/run_generation.py:401]: nightly version

File: examples/pytorch/text-generation/run_generation.py (hunk 1 above)
Comment (sywangyi) [reply examples/pytorch/text-generation/run_generation.py:253]: This is used for the first token generation.  It has no past_key_values in this case, but we still need to use the jit model which utilizes past_key_values in jit.trace input, so we have to pass the empty tensor here, because the passkey seq_len + input_seq = attention mask len.  And I have validated the usage, it works, same usage in huggingface optimum-intel.

File: examples/pytorch/text-generation/run_generation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply examples/pytorch/text-generation/run_generation.py:412]: As mentioned above - the examples should be simple demonstrations. As nightly builds are inherently unstable it makes this code brittle. We can either:
* Remove the torch 2.1.0 logic logic branch here
* Wait until 2.1.0 is released, remove the other branch and specify the torch version in the example requirements

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!

**Comment by sywangyi** [comment]
> Thanks for iterating!

Thanks @amyeroberts  could you help to merge the PR? I check the failed case in ci. it has nothing to do with my code.

**Comment by amyeroberts** [comment]
@sywangyi Looking at the CI output, it seems that the examples test run failed before the generation tests were run. Could you rebase from main to include any recent updates? I believe the accelerate errors should now be resolved.
